src/Bicep.Core/Semantics/Decorator.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Syntax;
+using Bicep.Core.TypeSystem;
+
+namespace Bicep.Core.Semantics
+{
+    public delegate void DecoratorValidator(
+        string decoratorName,
+        DecoratorSyntax decoratorSyntax,
+        TypeSymbol targetType,
+        ITypeManager typeManager,
+        IDiagnosticWriter diagnosticWriter);
+
+    public delegate ObjectSyntax? DecoratorEvaluator(
+        DecoratorSyntax decoratorSyntax,
+        TypeSymbol targetType,
+        ObjectSyntax? targetObject);
+
+    public class Decorator

```

Comment:
>Decorator [](start = 17, length = 9)

Is there any value in making decorators symbols? It might be helpful with some of the navigation features like hovers to avoid special casing...

Reply:
I thought about creating a new `DecoratorSymbol` type originally, but on second thought modelling them as `FunctionSymbol`s makes it easier to leverage all semantic checkers and LSP capablilities we have implemented for function calls (decorators are essentially higher order functions IMHO). Features like hovering is already working:
![image](https://user-images.githubusercontent.com/16367959/105772840-a9a54800-5f17-11eb-8ab7-ead98e8eb64d.png)

Do you mind clarifying what "to avoid special casing" means?

Reply:
Can ignore that comment if hovers are already working. I think the hover text will be confusing to our users. We should probably make it more obvious that it's a decorator. (Higher order functions are an advanced concept after all.)

Reply:
Agreed that we should make it more obvious that it's a decorator. Maybe showing descriptions below type signatures would help, but I feel like that's something to be implemented in another PR and we probably need to discuss it.